File: src/Board.java
import java.util.*;

public class Board {
    private int rows, cols;
    private char[][] grid;
    private Map<Character, Piece> pieces;
    private int exitRow, exitCol; // Posisi exit

    // ctor
    public Board(int rows, int cols, char[][] grid, Map<Character, Piece> pieces, int exitRow, int exitCol){
        this.rows = rows;
        this.cols = cols;
        this.grid = grid;
        this.pieces = pieces;
        this.exitRow = exitRow;
        this.exitCol = exitCol;
    }

    // Getter
    public int getRows(){
        return rows;
    }

    public int getCols(){
        return cols;
    }

    public char[][] getGrid(){
        return grid;
    }

    public Map<Character, Piece> getPieces(){
        return pieces;
    }

    public int getExitRow(){
        return exitRow;
    }

    public int getExitCol(){
        return exitCol;
    }

    public boolean isGoal(){
        Piece primary = null;

        for (Piece p : pieces.values()){
            if (p.isPrimary()){
                primary = p;
                break;
            }
        }

        if (primary == null){
            return false;
        }

        if (primary.isHorizontal()){
            return primary.getCol() + primary.getLength() - 1 == exitCol && primary.getRow() == exitRow;
        }else{
            return primary.getRow() + primary.getLength() - 1 == exitRow && primary.getCol() == exitCol;
        }
    }

    public Board copy(){
        char[][] newGrid = new char[rows][cols];

        for (int i = 0; i < rows; i++){
            newGrid[i] = Arrays.copyOf(grid[i], cols);
        }

        Map<Character, Piece> newPieces = new HashMap<>();

        for (Map.Entry<Character, Piece> entry : pieces.entrySet()){
            newPieces.put(entry.getKey(), entry.getValue().copy());
        }

        return new Board(rows, cols, newGrid, newPieces, exitRow, exitCol);
    }
}

File: src/lib/models/Board.ts
import { Piece } from "./Piece";

export class Board {
    constructor(
        private rows: number,
        private cols: number,
        private grid: string[][],
        private pieces: Map<string, Piece>,
        private exitRow: number,
        private exitCol: number
    ) {}

    getRows(): number {
        return this.rows;
    }

    getCols(): number {
        return this.cols;
    }

    getGrid(): string[][] {
        return this.grid;
    }

    getPieces(): Map<string, Piece> {
        return this.pieces;
    }

    getExitRow(): number {
        return this.exitRow;
    }

    getExitCol(): number {
        return this.exitCol;
    }

    isGoal(): boolean {
        let primary: Piece | undefined;
        for (const piece of this.pieces.values()) {
            if (piece.isPrimary) {
                primary = piece;
                break;
            }
        }

        if (!primary) return false;

        // Horizontal
        if (primary.isHorizontal) {
            // Right exit
            if (this.exitCol === this.cols) {
                return primary.row === this.exitRow && 
                       primary.col + primary.length - 1 === this.cols - 1;
            }
            // Left exit
            else if (this.exitCol === -1) {
                return primary.row === this.exitRow && primary.col === 0;
            }
        } 
        else {
            // Vertical
            // Bottom exit
            if (this.exitRow === this.rows) {
                return primary.col === this.exitCol && 
                       primary.row + primary.length - 1 === this.rows - 1;
            }
            // Top exit
            else if (this.exitRow === -1) {
                return primary.col === this.exitCol && primary.row === 0;
            }
        }
                
        return false;
    }

    copy(): Board {
        const newGrid = this.grid.map(row => [...row]);
        const newPieces = new Map<string, Piece>();
        for (const [key, piece] of this.pieces.entries()) {
            newPieces.set(key, piece.copy());
        }

        return new Board(this.rows, this.cols, newGrid, newPieces, this.exitRow, this.exitCol);
    }

    printBoard(): string {
        const output: string[] = [];
    
        // Top exit
        if (this.exitRow === -1) {
            const topLine = ' '.repeat(this.exitCol) + 'K';
            output.push(topLine);
        }
    
        // Grid rows
        for (let i = 0; i < this.rows; i++) {
            let line = "";
            if (this.exitCol === -1 && this.exitRow !== i) {
                line = " ";
            }
    
            // Left exit
            if (this.exitRow === i && this.exitCol === -1) {
                line += 'K';
            }
    
            for (let j = 0; j < this.cols; j++) {
                line += this.grid[i][j];
            }
    
            // Right exit
            if (this.exitRow === i && this.exitCol === this.cols) {
                line += 'K';
            }
    
            output.push(line);
        }
    
        // Bottom exit
        if (this.exitRow === this.rows) {
            const bottomLine = ' '.repeat(this.exitCol) + 'K';
            output.push(bottomLine);
        }
    
        return output.join('\n');
    }
}
